### ISSUE TITLE
[INFRA-2697] Unfork repos

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkins-infra/helpdesk#2272]
<p>As noted in <a href="https://github.com/cli/cli/issues/1467" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cli/cli/issues/1467</a> there are various repositories in @jenkinsci which are still formally considered forks of long-dormant personal repositories, even though they are in fact official. We should remove the linkage. <tt>jenkins-test-harness-htmlunit</tt> and <tt>plugin-pom</tt> I have noticed, but I am sure you could find others with a search.</p>

---
<details><summary><i>Originally reported by <a title="Jesse Glick" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a href="https://issues.jenkins.io/browse/INFRA-2697" target="_blank">Unfork repos</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Daniel Beck" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>
<li><b>status</b>: In Progress
<li><b>priority</b>: Major
<li><b>resolution</b>: Unresolved
<li><b>imported</b>: 2022/01/10
</ul></i>
</details>

**Comment by offa** [comment]
Could you consider unforking of these repos please?

- https://github.com/jenkinsci/git-plugin
- https://github.com/jenkinsci/webhook-step-plugin
- https://github.com/jenkinsci/parameterized-scheduler-plugin

**Comment by timja** [comment]
cc @MarkEWaite, @jmMeessen, @res0nance ^^

**Comment by basil** [comment]
I've got one of these too - https://github.com/jenkinsci/jakarta-mail-api-plugin - but I have no idea how I would go about unforking it.

**Comment by NotMyFault** [comment]
If nothing changed over the past couple of years, this is still done manually by submitting a [request](https://support.github.com/request) to GitHub.
Looks like you need to be an org admin for that.

**Comment by MarkEWaite** [comment]
Based on https://stackoverflow.com/questions/38831301/how-to-un-fork-the-github-repository/66470086#66470086 I was able to use their virtual assistant to submit a request to unfork the elastic axis plugin and the git plugin.  I hit the rate limit when I tried to unfork the node label parameter plugin

**Comment by timja** [comment]
I used the virtual assistant recently for one, plugin-pom I think.

**Comment by MarkEWaite** [comment]
> I used the virtual assistant recently for one, plugin-pom I think. It was done fairly quickly

All four of my requests are now complete.  The request for git plugin is resolved.

**Comment by NotMyFault** [comment]
Btw, is there a policy for the forking process, because https://github.com/jenkinsci/ecu-test-execution-plugin was just forked ~10h ago and may be considered as unforking candidate sooner or later 🤔

**Comment by timja** [comment]
see https://github.com/jenkins-infra/repository-permissions-updater/issues/2679#issuecomment-1206357817

(wording could probably be updated as there's now an official process from GitHub side on it)

**Comment by basil** [comment]
I successfully used the virtual assistant to unfork https://github.com/jenkinsci/jakarta-mail-api-plugin - thanks!

**Comment by basil** [comment]
Here are a few more I found that need to be dealt with:

https://github.com/jenkinsci/active-choices-plugin
https://github.com/jenkinsci/build-pipeline-plugin
https://github.com/jenkinsci/conditional-buildstep-plugin
https://github.com/jenkinsci/docker-java-api-plugin
https://github.com/jenkinsci/ecu-test-execution-plugin
https://github.com/jenkinsci/git-parameter-plugin
https://github.com/jenkinsci/job-dsl-plugin
https://github.com/jenkinsci/jquery-detached
https://github.com/jenkinsci/kubernetes-credentials-plugin
https://github.com/jenkinsci/mapdb-api-plugin
https://github.com/jenkinsci/parameterized-scheduler-plugin
https://github.com/jenkinsci/pipeline-graph-analysis-plugin
https://github.com/jenkinsci/publish-over-plugin
https://github.com/jenkinsci/publish-over-ssh-plugin
https://github.com/jenkinsci/run-condition-plugin
https://github.com/jenkinsci/snakeyaml-api-plugin
https://github.com/jenkinsci/tikal-multijob-plugin
https://github.com/jenkinsci/webhook-step-plugin

**Comment by NotMyFault** [comment]
I submited a request for repositories I can manage 👍🏻

**Comment by basil** [comment]
Just browsing around a few popular plugins.

**Comment by timja** [comment]
There's a report here that has all plugins in this situation:
https://reports.jenkins.io/github-jenkinsci-fork-report.json

**Comment by timja** [comment]
I've created a request for the snakeyaml one

**Comment by NotMyFault** [comment]
> There's a report here that has all plugins in this situation: https://reports.jenkins.io/github-jenkinsci-fork-report.json

If we want to continue detaching forks, I'd recommend that one person submits all eligible repositories on behalf of the jenkinsci umbrella organization.
I don't think we necessarily need to create one ticket per person, if we can and want to get all done in one row.

**Comment by timja** [comment]
One of the questions is 'have you committed to the repository'

I don't think one person will meet that criteria, although possibly they would let an organisation owner do it.

**Comment by NotMyFault** [comment]
> One of the questions is 'have you committed to the repository'

That was no blocker for a repo I detached an hour ago.

> although possibly they would let an organisation owner do it.

I just asked them on my ticket, 'cause I'd rather not want to submit 770 tickets.
Gonna add it up when they respond.

**Comment by NotMyFault** [comment]
> Since high volume requests consume a non-trivial amount of resources (and our support tooling isn't equipped to handle more than a few forks at a time), this has been escalated to another team for further review.

> While we don't yet have an ETA, we'll be sure to be in touch with next steps, or to keep you apprised of any news.

Still ongoing.

**Comment by NotMyFault** [comment]
GH evaluated the feasibility internally, but given unforking is a manual resource and time intensive process, GH is unable to unfork the vast amount of forks we host.

However, they offered us to change all fork network owners to the jenkinsci organization, if desired.

**Comment by timja** [comment]
Should be fine to do that, cc @daniel-beck

**Comment by NotMyFault** [comment]
I sent you the email to authorize the process, Tim.

**Comment by daniel-beck** [comment]
> However, they offered us to change all fork network owners to the jenkinsci organization, if desired.

This would be wrong for some repos, so we would provide either a list of repos to apply this to, or a list of repos to _not_ apply this to. Otherwise https://github.com/jenkinsci/jgit https://github.com/jenkinsci/rubywm https://github.com/jenkinsci/plexus-utils https://github.com/jenkinsci/jna https://github.com/jenkinsci/jmdns-fork would probably upset some folks.

**Comment by daniel-beck** [comment]
> There's a report here that has all plugins in this situation:
> https://reports.jenkins.io/github-jenkinsci-fork-report.json

FTR https://www.jenkins.io/doc/developer/publishing/source-code-hosting/forks/

**Comment by timja** [comment]
We could delete those forks? I don't think any of them are used now

**Comment by daniel-beck** [comment]
> We could delete those forks? I don't think any of them are used now

These are a handful of examples I found when browing through the repos list. Is that something you can confidently say about every fork that should be a fork?

**Comment by NotMyFault** [comment]
> > However, they offered us to change all fork network owners to the jenkinsci organization, if desired.
> 
> This would be wrong for some repos, so we would provide either a list of repos to apply this to, or a list of repos to _not_ apply this to. Otherwise https://github.com/jenkinsci/jgit https://github.com/jenkinsci/rubywm https://github.com/jenkinsci/plexus-utils https://github.com/jenkinsci/jna https://github.com/jenkinsci/jmdns-fork would probably upset some folks.

I started writing down a non exhaustive list of forks that should remain as is, given those are legit forks outside the plugin hosting process: https://gist.github.com/NotMyFault/dfec71b094419f45145d2c187636b6b1

**Comment by daniel-beck** [comment]
> non exhaustive list of forks

Can you do this as a list of all forks we have, and label them "actual fork" and "hosting fork"? (FWIW I'd use a 2 column Google Doc for this.) That way it's easy to confirm all repos were looked at, and an explicit decision to unfork/root (or not) was made. Especially interesting also for weird edge cases (we also have some non-plugin forks of vaguely Jenkins related repos of contributors…)

**Comment by timja** [comment]
GitHub have not offered to process a list from what I understand they would do all of them or small numbers.

If we don’t want something unforked and we want to do this process we need to delete the repo, which could mean moving it to a temp account and back

**Comment by daniel-beck** [comment]
> If we don’t want something unforked and we want to do this process we need to delete the repo, which could mean moving it to a temp account and back

Right, but that still means we should have a complete list to minimize the risk that we have forked another project and suddenly become root of their network.

**Comment by NotMyFault** [comment]
> GitHub have not offered to process a list from what I understand they would do all of them or small numbers.

Right, GH's offer covers all forks.

> have a complete list

Based on the fork report, I generated https://docs.google.com/spreadsheets/d/197A5yzVCfBsXkRF3Uh9JbSUKD2kU7q-z6O8rNxLdhNE/edit#gid=0, which highlights repositories that should remain as is and we would need to move to a temp account (or possibly delete, if no commits have been made and the repository brings in nothing of use and and value).

**Comment by daniel-beck** [comment]
> highlights repositories that should remain as is and we would need to move to a temp account

Right, and now explicitly mark those that should have their network root in jenkinsci. TBH I'd be pretty surprised if that's trivially decidable for everything not currently marked to remain a fork.

**Comment by timja** [comment]
I've reviewed all of the repositories not containing `-plugin` and added a couple more.

> Right, and now explicitly mark those that should have their network root in jenkinsci. TBH I'd be pretty surprised if that's trivially decidable for everything not currently marked to remain a fork.

I don't quite understand. Shouldn't all the plugins have their network root as jenkinsci?

**Comment by daniel-beck** [comment]
> I don't quite understand. Shouldn't all the plugins have their network root as jenkinsci?

All all repos not marked to remain a fork plugins? No, plenty of counter-examples.

Are all `-plugin` Jenkins plugins? No, see the already marked Maven plugin.

**Comment by NotMyFault** [comment]
I seperated the list across different sheets on the spreadsheet, but I also didn't find anything new reading over it.

Given that changing fork networks is a simple procedure on GH's end and we discover a fork we want to restore in the future, we can always go ahead and do so and prepare the few forks giving GH the go.

**Comment by daniel-beck** [comment]
The point of my recent messages is that we need to be able to distinguish "action should be taken" and "nobody has looked at this so we don't know yet".

I know of at least one repo currently marked for rooting in this doc that shouldn't be because it's an actively maintained open source project unrelated to Jenkins that we forked. This alone tells me that this was not done properly.

**Comment by hashar** [comment]
> Based on the fork report, I generated https://docs.google.com/spreadsheets/d/197A5yzVCfBsXkRF3Uh9JbSUKD2kU7q-z6O8rNxLdhNE/edit#gid=0, which highlights repositories that should remain as is and we would need to move to a temp account (or possibly delete, if no commits have been made and the repository brings in nothing of use and and value).

As the maintainer of https://github.com/jenkinsci/gearman-plugin.git I would very welcome it to become its own fork network (in case you are seeking feedback).

The plugin development was originally done on OpenStack Gerrit instance with a mirror set to https://github.com/openstack-infra/gearman-plugin . I think I forked it under the jenkinsci organization for convenience but there is really no need to keep that fork relation since the parent was merely a mirror, it had no issue tracker and the PR were automatically rejected.  When proposing a PR to `jenkinsci/gearman-plugin` GitHub web UI defaults to fill the PR against the obsolete parent `openstack-infra/gearman-plugin` which is a bit confusing.

Thanks for leading that effort with GitHub!

**Comment by daniel-beck** [comment]
@hashar Note that GH said they cannot disconnect our stuff in batch (https://github.com/jenkins-infra/helpdesk/issues/2272#issuecomment-1216867119), but instead offered to make the jenkinsci fork the new root of the network.
